src/descformats/base.py

```diff
@@ -0,0 +1,284 @@
+    machinery of the PipelineStage class to open an input our output
+    named by a tag.
+
+    """
+
+    @classmethod
+    def _open(cls, path, mode, **kwargs):
+        """
+        Open a data file.  The base implementation of this function just
+        opens and returns a standard python file object.
+
+        Subclasses can override to either open files using different openers
+        (like fitsio.FITS), or, for more specific data types, return an
+        instance of the class itself to use as an intermediary for the file.
+
+        """
+        return open(path, mode, encoding='utf-8')
+
+    @classmethod
+    def _close(cls, fileObj, **kwargs):
+        fileObj.close()
+
+
+
+class TableHandle(DataHandle):
+    """DataHandle for single tables of data
+    """
+    suffix = "fits"
+
+    @classmethod
+    def _open(cls, path, mode, **kwargs):
+        """Open and return the associated file
+
+        Notes
+        -----
+        This will simply open the file and return a file-like object to the caller.
+        It will not read or cache the data
+        """
+        return tables_io.io.io_open(path, **kwargs)  #pylint: disable=no-member
+
+    @classmethod
+    def _read(cls, path, **kwargs):
+        """Read and return the data from the associated file """
+        return tables_io.read(path, **kwargs)
+
+    @classmethod
+    def _write(cls, data, path, **kwargs):
+        """Write the data to the associatied file """
+        return tables_io.write(data, path, **kwargs)
+
+    @classmethod
+    def _size(cls, path, **kwargs):
+        return tables_io.io.getInputDataLengthHdf5(path, **kwargs)
+
+    @classmethod
+    def _iterator(cls, path, **kwargs):
+        """Iterate over the data"""
+        return tables_io.iteratorNative(path, **kwargs)
+
+
+class Hdf5Handle(TableHandle):
+    """A data file in the HDF5 format.
+    Using these files requires the h5py package, which in turn
+    requires an HDF5 library installation.
+    """
+    suffix = 'hdf5'
+    format = "http://edamontology.org/format_3590"    
+    required_datasets = []
+
+    @classmethod
+    def _open(cls, path, mode, **kwargs):
+        # Suppress a warning that h5py always displays
+        # on import
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            import h5py  # pylint: disable=import-outside-toplevel
+        # Return an open h5py File
+        return h5py.File(path, mode, **kwargs)
+
+    def validate(self):
+        if self.data is not None:
+            return
+        if self.fileObj is None:
+            self.open(mode='r')
+        missing = [name for name in self.required_datasets if name not in self.fileObj]
+        if missing:
+            text = "\n".join(missing)
+            raise FileValidationError(f"These data sets are missing from HDF file {self.path}:\n{text}")
+
+    @classmethod
+    def _write_chunk(cls, data, fileObj, groups, start, end, **kwargs):
```

Comment:
Should there be a corresponding read_chunk? Or is that covered in the iterator method?

Reply:
That is covered in the iterator(), but it might be useful to be able to read a single chunk...

Reply:
Could leave this for now until we need it.